compiler/il/OMRNode.cpp

```diff
@@ -2357,7 +2357,101 @@ OMR::Node::isNotCollected()
    return false;
    }
 
+/**
+ * \brief
+ *    Check if the node is an internal pointer.
+ *
+ * \return
+ *    Return true if the node is an internal pointer, false otherwise.
+ *
+ * \note
+ *    This query currently only works on opcodes that can have a pinning array pointer,
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   if (self()->getOpCode().hasPinningArrayPointer())
+      return false;
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collectable.
+ *
+ * \return
+ *    Return true if the node is collectable, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from a collectable reference, i.e. anything that is an address to a collected object.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (!curNode->getType().isAddress())
+      return false;
+
+   while (curNode)
+      {
+      if (curNode->isInternalPointer())
+         return true;
 
+      if (curNode->getOpCode().hasSymbolReference())
+         {
+         TR::ILOpCode op = curNode->getOpCode();
+         TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+         // Current implementation treats all object references collectable, and also
+         // assumes that the return of an acall* is an object reference.
+         if (op.isNew() ||
+             (op.getDataType() == TR::Address && op.isCall()) ||
+             symbol->isInternalPointer() ||
+             symbol->isCollectedReference())
+             return true;
+         }
+
+      if (curNode->getNumChildren() >= 1)
+         {
+         TR::ILOpCode childOp = curNode->getFirstChild()->getOpCode();
+         TR::ILOpCodes ops = childOp.getOpCodeValue();
+         switch (ops)
+            {
+            case TR::aload:
+            case TR::aloadi:
+            case TR::aRegLoad:
+            case TR::areturn:
+            case TR::acall:
+            case TR::acalli:
+            case TR::compressedRefs:
+            case TR::New:
+            case TR::newarray:
+            case TR::anewarray:
+            case TR::variableNew:
+            case TR::variableNewArray:
+            case TR::multianewarray:
+            case TR::aiadd:
+            case TR::aladd:
+               curNode = curNode->getFirstChild();
+               break;
+            case TR::aternary:
+               {
+               bool isFirstChildCollected = curNode->getFirstChild()->computeIsCollectedReference();
+               bool isSecondChildCollected = curNode->getSecondChild()->computeIsCollectedReference();
+               TR_ASSERT(isFirstChildCollected != isSecondChildCollected, "Computation does not support address node from aternary with one child collectable the other uncollectable, node is " POINTER_PRINTF_FORMAT, self());
```

Comment:
I think this assert should say that 'aternary' does not support having one child collected and the other not collected. It won't work in the codegen and is just not supported so I would state that rather than the computation doesn't support it - the compiler in general doesn't support it.